### ISSUE TITLE
-Wpointer-bool-conversion/-Waddress misses an obvious case

### ISSUE THREAD

**Issue by r-barnes** [llvm/llvm-project#82512]
The following pattern is often used to evaluate an expensive function only once, cache its result, and return that on subsequent calls.
```
int foo() {
    static bool is_true = [](){ return expensive_helper_function; };

    return is_true;
}
```
But this code is subtly wrong. It should be
```
static bool is_true = [](){ return expensive_helper_function; }(); // Note the parens!
```
In the bad case it looks as though the lambda is being evaluated to a pointer(?) and this is being evaluated always to `true` when it is assigned to `is_true`.

It feels as though `-Wpointer-bool-conversion` should flag this, but it does not ([godbolt](https://godbolt.org/z/v9qhze9e3)).

**Comment by MatzeB** [comment]
I think I filed/proposed this in an issue this years ago  #25841 ...

**Comment by MatzeB** [comment]
FWIW this seems to be covered by clang-tidy https://clang.llvm.org/extra/clang-tidy/checks/readability/implicit-bool-conversion.html but unfortunately this rule also applies to other forms of conversions like `int -> bool` which may make it less suitable to enable in all situations...

**Comment by shafik** [comment]
So we end up here:

https://github.com/llvm/llvm-project/blob/004c1972b4585fe8051814ceb6c6cdbf3cb62290/clang/lib/Sema/SemaChecking.cpp#L15628-L15632

and then we bail out here:

https://github.com/llvm/llvm-project/blob/004c1972b4585fe8051814ceb6c6cdbf3cb62290/clang/lib/Sema/SemaChecking.cpp#L16463-L16465

I think what we need is a extra check in `Sema::DiagnoseAlwaysNonNullPointer(...)`

check for `CXXMemberCallExpr` and then if `getObjectType()` gives us a `RecordType` then check `isLambda()` and diagnose.

@AaronBallman seems like a good first issue.

**Comment by AaronBallman** [comment]
This does seem like a reasonable first issue and those steps are what I'd recommend as well. I don't think there are any surprises lurking in here that would trip someone up.

**Comment by vinayakdsci** [comment]
@AaronBallman  Hello! I would love to work on this issue. Could it be assigned to me?

**Comment by EugeneZelenko** [comment]
@vinayakdsci: Just create pull request and mention it on this page.

**Comment by vinayakdsci** [comment]
@AaronBallman I was trying to follow the instructions provided by Shafik above, and I seem to be getting an assertion failure `DiagnosticsEngine::ArgumentKind clang::Diagnostic::getArgKind(unsigned int) const: Assertion Idx < getNumArgs() && "Argument index out of range!"' failed`.

This is triggered when I try to emit a diagnostics with DiagID `diag::warn_impcast_pointer_to_bool`.

Do you have any ideas on how to fix this? I am unable to get it to work, and have tried a number of ways to avoid this assertion failure.

This diagnostic I was able to locate in the diagnostic tablegen file `DiagnosticSemaKinds.td`.

Thanks!

**Comment by dwblaikie** [comment]
you might not be passing the right number of/enough parameters to the diagnostic? In the .td file you'll see the diagnostic text uses a number of placeholders (%0, etc I think) and you'll need to supply just as many to the diagnostic - mabye check another use of that diagnostic to see what/how the parameters are passed?

**Comment by vinayakdsci** [comment]
@dwblaikie thanks for the advice, I was able to fix the errors. However, after the fixes, some of the existing tests are failing.
One is in `clang/test/CXX/drs/dr18xx.cpp` where a lambda expression is assigned to bool and three in `CXX/expr/expr.prim/expr.prim.lamda/blocks.mm` where a struct inherits from a lambda function. While I think that the former can fixed with a warning, I am not sure if it would be alright to change any tests in `expr.prim.lambda/`?

How do you think I should proceed? Should I open a PR and wait for review, and then fix any errors?

Thanks!
